### ISSUE TITLE
`manual_map` does not consider the borrowing relationship

### ISSUE THREAD

**Issue by timongh** [rust-lang/rust-clippy#9407]
### Summary

In the example below, the suggestion given by clippy is wrong. Because of the conflicting borrows in the Scrutinee and Arm of the match.

### Lint Name

manual_map

### Reproducer

I tried this code (try it in [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=25cfdb210fcdf97a3e9b623701f08ef9)):

```rust
struct A(i32);

impl A {
    fn val(&self) -> Option<&i32> {
        Some(&self.0)
    }
    
    fn do_something(&mut self) -> &'static str {
        "result"
    }
}

let mut a = A(1);
let _ = match a.val() {
    None => None,
    Some(_) => Some(a.do_something()),
};
```

I saw this happen:

```
warning: manual implementation of `Option::map`
  --> src/main.rs:15:13
   |
15 |       let _ = match a.val() {
   |  _____________^
16 | |         None => None,
17 | |         Some(_) => Some(a.do_something()),
18 | |     };
   | |_____^ help: try this: `a.val().map(|_| a.do_something())`
   |
   = note: `#[warn(clippy::manual_map)]` on by default
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_map
```

I expected to see this happen:
